File: Week_02/94_binary_tree_inorder_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// Recursion
class Solution0 {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();

        helper(root, res);

        return res;
    }

    private void helper(TreeNode root, List<Integer> result){
        if(root == null) return;
        if(root.left != null) helper(root.left, result);
        result.add(root.val);
        if(root.right != null) helper(root.right, result);
    }
}

// Stack
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        Deque<TreeNode> stack =  new LinkedList<TreeNode>();
        TreeNode node = root;

        while(!stack.isEmpty() || node != null) {
            while (node != null) {
                stack.push(node);
                node = node.left;
            }

            node = stack.pop();
            res.add(node.val);
            node = node.right;

        }
        return res;
    }
}

File: Week_03/python/tree_traverse/94_binary_tree_inorder_traversal.py
# Definition for a binary tree node.
from collections import deque
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        return self.inorder(root)

    def inorder(self, root: TreeNode):
        if not root:
            return []

        return self.inorder(root.left) + [root.val] + self.inorder(root.right)


class Solution2:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        stack = deque()
        ans = []

        node = root

        while len(stack) > 0 or node:
            while node:
                stack.append(node)
                node = node.left

            node = stack.pop()
            ans.append(node.val)
            node = node.right

        return ans
